File: leetcode/kth-smallest-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        return inorderTraversal(root).get(k-1);
    }
    
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> order = new ArrayList<>();
        
        if(root == null)
            return order;
        
        if(root.left != null)
            order.addAll(inorderTraversal(root.left));
        
        order.add(root.val);
        
        if(root.right != null)
            order.addAll(inorderTraversal(root.right));
        
        return order;
    }
}

File: leetcode/kth-smallest-in-a-bst.py
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        """
        In-order traversal, return the kth element in the traversal


        """
        def dfs(node: TreeNode):
            nonlocal k
            if not node:
                return None

            left = dfs(node.left)
            if left: 
                return left

            k -= 1
            print(f"k: {k}, node.val: {node.val}")
            if k == 0:
                return node

            right = dfs(node.right)
            if right:
                return right

        kth_smallest = dfs(root)
        return None if not kth_smallest else kth_smallest.val

        
